src/actions/FreeRun.ts

```diff
@@ -0,0 +1,300 @@
+import {
+  cliExecute,
+  mallPrice,
+  myTurncount,
+  restoreMp,
+  retrieveItem,
+  visitUrl,
+} from "kolmafia";
+
+import { Macro } from "../combat";
+import {
+  ensureEffect,
+  getFoldGroup,
+  getSongCount,
+  getSongLimit,
+  have,
+} from "../lib";
+import { Requirement } from "../maximize";
+import { get } from "../property";
+import { $effect, $familiar, $item, $items, $skill } from "../template-string";
+import {
+  ActionSource,
+  findActionSource,
+  FindActionSourceConstraints,
+} from "./ActionSource";
+import * as Bandersnatch from "../resources/2009/Bandersnatch";
+import * as AsdonMartin from "../resources/2017/AsdonMartin";
+
+// Value of _lastCombatStarted the last time we updated scrapbook charges.
+let scrapbookChargesLastUpdated = get("_lastCombatStarted");
+
+// Free unlimited source every 30 turns.
+// Does not work on special monsters so needs a backup, see tryFindFreeRun.
+const asdonMartinSource: ActionSource = new ActionSource(
+  $skill`Asdon Martin: Spring-Loaded Front Bumper`,
+  () => {
+    if (!AsdonMartin.installed()) return 0;
+    const banishes = get("banishedMonsters").split(":");
+    const bumperIndex = banishes
+      .map((string) => string.toLowerCase())
+      .indexOf("spring-loaded front bumper");
+    if (bumperIndex === -1) return 1;
+    return myTurncount() - parseInt(banishes[bumperIndex + 1]) > 30 ? 1 : 0;
+  },
+  Macro.skill($skill`Asdon Martin: Spring-Loaded Front Bumper`),
+  {
+    preparation: () => AsdonMartin.fillTo(50),
+  }
+);
+
+const freeRunSources: ActionSource[] = [
+  // Free limited sources
+  new ActionSource(
+    $familiar`Frumious Bandersnatch`,
+    () =>
+      have($familiar`Frumious Bandersnatch`) &&
+      (have($effect`Ode to Booze`) || getSongCount() < getSongLimit()) &&
+      Bandersnatch.getRemainingRunaways() > 0
+        ? 0
```

Comment:
Right, I never had this return the potential, so it should for now just return Bandersnatch.getRemainingRunaways(). Same for boots.